File: tests/r_parity/42_nbreg.py
"""StatsPAI negative binomial parity (Python side) -- Module 42."""
from __future__ import annotations
import numpy as np, pandas as pd, statspai as sp
from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "42_nbreg"

def make_data(n=600, seed=PARITY_SEED):
    rng = np.random.default_rng(seed)
    x1 = rng.normal(0, 1, n)
    x2 = rng.binomial(1, 0.5, n).astype(float)
    mu = np.exp(0.8 + 0.5 * x1 - 0.4 * x2)
    # NB with dispersion alpha = 1.0
    alpha = 1.0
    p = mu / (mu + 1.0 / alpha)
    n_param = 1.0 / alpha
    y = rng.negative_binomial(n=n_param, p=1 - p)
    return pd.DataFrame({"y": y.astype(float), "x1": x1, "x2": x2})

def main():
    df = make_data()
    dump_csv(df, MODULE)
    res = sp.nbreg(formula="y ~ x1 + x2", data=df)
    rows = []
    for nm, label in [("_cons", "intercept"), ("x1", "x1"), ("x2", "x2")]:
        if nm in res.params.index:
            rows.append(ParityRecord(MODULE, "py",
                f"beta_{label}",
                estimate=float(res.params[nm]),
                se=float(res.std_errors[nm]),
                n=int(len(df))))
    write_results(MODULE, "py", rows, extra={"dispersion": "mean", "engine": "statsmodels"})

if __name__ == "__main__":
    main()

File: tests/r_parity/42_nbreg.R
# StatsPAI negative binomial parity (R side) -- Module 42.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) dirname(normalizePath(sub("^--file=", "", .file_arg[1]))) else getwd()
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({ library(MASS) })

MODULE <- "42_nbreg"
df <- read_csv_strict(MODULE)

fit <- MASS::glm.nb(y ~ x1 + x2, data = df)
co <- coef(fit); se <- sqrt(diag(vcov(fit)))

rows <- list(
  parity_row(MODULE, "beta_intercept", estimate = unname(co["(Intercept)"]), se = unname(se["(Intercept)"]), n = nrow(df)),
  parity_row(MODULE, "beta_x1",        estimate = unname(co["x1"]),          se = unname(se["x1"]),          n = nrow(df)),
  parity_row(MODULE, "beta_x2",        estimate = unname(co["x2"]),          se = unname(se["x2"]),          n = nrow(df))
)

write_results(MODULE, rows, extra = list(package = "MASS::glm.nb"))
